File: sdk/typescript/src/internal/package-graph.ts
import { createHash } from "node:crypto"

import {
  canonicalizeJsonValue,
  parseJson,
  type JsonObject,
  type JsonValue,
} from "./jsoncanon"

export const PACKAGE_GRAPH_FORMAT_VERSION = 0 as const

const maxPackageGraphSizeBytes = 16777216
const maxPackageNameBytes = 214
const maxPackageVersionBytes = 255
const maxPackagePathComponentBytes = 255
const maxMountedPackagePathBytes = 4096
const localPackageViewKeyDomain = "helmr.local-package-view.v0"
const programMountPath = "/opt/helmr/program"
const dependencyMountPath = "/opt/helmr/program/node_modules"
const packageNamePattern =
  /^(?:[a-z0-9][a-z0-9._~-]*|@[a-z0-9][a-z0-9._~-]*\/[a-z0-9][a-z0-9._~-]*)$/
const sha256DigestPattern = /^sha256:[0-9a-f]{64}$/
const sha512IntegrityPattern = /^sha512-[A-Za-z0-9+/]{86}==$/
const textEncoder = new TextEncoder()

export type PackageKind = "local" | "registry"
export type PackageRelationship = "production" | "optional" | "peer"

export interface LocalPackage {
  readonly manifestDigest: string
  readonly name: string | null
  readonly path: string
  readonly version: string | null
  readonly viewKey: string | null
}

export interface RegistryPackage {
  readonly installPath: string
  readonly integrity: string
  readonly name: string
  readonly version: string
}

export type PackageEndpoint =
  | Readonly<{ kind: "local"; path: string }>
  | Readonly<{ installPath: string; kind: "registry" }>

export interface PackageResolution {
  readonly dependency: string
  readonly from: PackageEndpoint
  readonly relationship: PackageRelationship
  readonly to: PackageEndpoint
}

export interface PackageGraph {
  readonly formatVersion: 0
  readonly localPackages: readonly LocalPackage[]
  readonly registryPackages: readonly RegistryPackage[]
  readonly resolutions: readonly PackageResolution[]
}

export function parsePackageGraph(raw: string | Uint8Array): PackageGraph {
  const input = typeof raw === "string" ? textEncoder.encode(raw) : raw
  if (input.length === 0 || input.length > maxPackageGraphSizeBytes) {
    throw new Error(`package graph size must be between 1 and ${maxPackageGraphSizeBytes} bytes`)
  }
  const value = parseJson(raw)
  const canonical = canonicalizeJsonValue(value)
  if (!bytesEqual(input, canonical)) {
    throw new Error("package graph is not RFC 8785 canonical JSON")
  }
  return validatePackageGraphValue(value)
}

export function canonicalPackageGraph(graph: PackageGraph): Uint8Array {
  validatePackageGraph(graph)
  const canonical = canonicalizeJsonValue(graph as unknown as JsonValue)
  if (canonical.length > maxPackageGraphSizeBytes) {
    throw new Error(`package graph size must be at most ${maxPackageGraphSizeBytes} bytes`)
  }
  return canonical
}

export function validatePackageGraph(graph: PackageGraph): void {
  canonicalizeJsonValue(graph as unknown as JsonValue)
  validatePackageGraphValue(graph as unknown as JsonValue)
}

function validatePackageGraphValue(value: JsonValue): PackageGraph {
  const root = requireObject(value, "package graph")
  requireKeys(
    root,
    ["formatVersion", "localPackages", "registryPackages", "resolutions"],
    "package graph",
  )
  if (root["formatVersion"] !== PACKAGE_GRAPH_FORMAT_VERSION) {
    throw new Error(`package graph formatVersion must be ${PACKAGE_GRAPH_FORMAT_VERSION}`)
  }

  const localValues = requireArray(root["localPackages"], "package graph localPackages")
  const registryValues = requireArray(root["registryPackages"], "package graph registryPackages")
  const resolutionValues = requireArray(root["resolutions"], "package graph resolutions")
  const localPackages = localValues.map((item, position) => parseLocalPackage(item, position))
  const registryPackages = registryValues.map((item, position) =>
    parseRegistryPackage(item, position),
  )
  if (localPackages.length === 0 || localPackages[0]!.path !== ".") {
    throw new Error("package graph localPackages must begin with exactly one root record")
  }

  const locals = new Map<string, LocalPackage>()
  const localNames = new Set<string>()
  const viewKeys = new Set<string>()
  for (let position = 0; position < localPackages.length; position++) {
    const local = localPackages[position]!
    validateLocalPackage(local, position === 0)
    if (locals.has(local.path)) {
      throw new Error(`package graph localPackages contains duplicate path ${JSON.stringify(local.path)}`)
    }
    locals.set(local.path, local)
    if (local.name !== null) {
      if (localNames.has(local.name)) {
        throw new Error(`package graph localPackages contains ambiguous name ${JSON.stringify(local.name)}`)
      }
      localNames.add(local.name)
    }
    if (local.viewKey !== null) {
      if (viewKeys.has(local.viewKey)) {
        throw new Error(`package graph localPackages contains colliding viewKey ${JSON.stringify(local.viewKey)}`)
      }
      viewKeys.add(local.viewKey)
    }
    if (position > 1) {
      const previous = localPackages[position - 1]!.path
      if (compareUtf8(previous, local.path) >= 0) {
        throw new Error(`package graph localPackages are not in canonical path order at position ${position}`)
      }
    }
    for (let separator = local.path.indexOf("/"); separator >= 0; ) {
      const ancestor = local.path.slice(0, separator)
      if (locals.has(ancestor)) {
        throw new Error(`package graph non-root local paths ${JSON.stringify(ancestor)} and ${JSON.stringify(local.path)} overlap`)
      }
      separator = local.path.indexOf("/", separator + 1)
    }
  }

  const registries = new Set<string>()
  for (let position = 0; position < registryPackages.length; position++) {
    const registry = registryPackages[position]!
    validateRegistryPackage(registry)
    if (registries.has(registry.installPath)) {
      throw new Error(`package graph registryPackages contains duplicate installPath ${JSON.stringify(registry.installPath)}`)
    }
    registries.add(registry.installPath)
    if (
      position > 0 &&
      compareUtf8(registryPackages[position - 1]!.installPath, registry.installPath) >= 0
    ) {
      throw new Error(`package graph registryPackages are not in canonical installPath order at position ${position}`)
    }
  }

  const resolutions = resolutionValues.map((item, position) =>
    parseResolution(item, position, locals, registries),
  )
  for (let position = 1; position < resolutions.length; position++) {
    if (compareResolutions(resolutions[position - 1]!, resolutions[position]!) >= 0) {
      throw new Error(`package graph resolutions are not in canonical order at position ${position}`)
    }
  }

  return {
    formatVersion: PACKAGE_GRAPH_FORMAT_VERSION,
    localPackages,
    registryPackages,
    resolutions,
  }
}

function parseLocalPackage(value: JsonValue, position: number): LocalPackage {
  const local = requireObject(value, `package graph localPackages[${position}]`)
  requireKeys(
    local,
    ["manifestDigest", "name", "path", "version", "viewKey"],
    `package graph localPackages[${position}]`,
  )
  return {
    manifestDigest: requireDigest(
      local["manifestDigest"],
      `package graph localPackages[${position}].manifestDigest`,
    ),
    name: requireNullableString(local["name"], `package graph localPackages[${position}].name`),
    path: requireString(local["path"], `package graph localPackages[${position}].path`),
    version: requireNullableString(
      local["version"],
      `package graph localPackages[${position}].version`,
    ),
    viewKey: requireNullableString(
      local["viewKey"],
      `package graph localPackages[${position}].viewKey`,
    ),
  }
}

function parseRegistryPackage(value: JsonValue, position: number): RegistryPackage {
  const registry = requireObject(value, `package graph registryPackages[${position}]`)
  requireKeys(
    registry,
    ["installPath", "integrity", "name", "version"],
    `package graph registryPackages[${position}]`,
  )
  return {
    installPath: requireString(
      registry["installPath"],
      `package graph registryPackages[${position}].installPath`,
    ),
    integrity: requireString(
      registry["integrity"],
      `package graph registryPackages[${position}].integrity`,
    ),
    name: requireString(registry["name"], `package graph registryPackages[${position}].name`),
    version: requireString(
      registry["version"],
      `package graph registryPackages[${position}].version`,
    ),
  }
}

function parseResolution(
  value: JsonValue,
  position: number,
  locals: ReadonlyMap<string, LocalPackage>,
  registries: ReadonlySet<string>,
): PackageResolution {
  const resolution = requireObject(value, `package graph resolutions[${position}]`)
  requireKeys(
    resolution,
    ["dependency", "from", "relationship", "to"],
    `package graph resolutions[${position}]`,
  )
  const dependency = requireString(
    resolution["dependency"],
    `package graph resolutions[${position}].dependency`,
  )
  validatePackageName(dependency, `package graph resolutions[${position}].dependency`)
  const relationship = resolution["relationship"]
  if (relationship !== "production" && relationship !== "optional" && relationship !== "peer") {
    throw new Error(`package graph resolutions[${position}].relationship is unsupported`)
  }
  const from = parseEndpoint(
    resolution["from"],
    `package graph resolutions[${position}].from`,
    locals,
    registries,
  )
  const to = parseEndpoint(
    resolution["to"],
    `package graph resolutions[${position}].to`,
    locals,
    registries,
  )
  if (to.kind === "local") {
    const local = locals.get(to.path)!
    if (local.name === null) {
      throw new Error(`package graph local target ${JSON.stringify(to.path)} has no name`)
    }
    if (dependency !== local.name) {
      throw new Error(`package graph dependency ${JSON.stringify(dependency)} does not equal local target name ${JSON.stringify(local.name)}`)
    }
  }
  return { dependency, from, relationship, to }
}

function parseEndpoint(
  value: JsonValue | undefined,
  label: string,
  locals: ReadonlyMap<string, LocalPackage>,
  registries: ReadonlySet<string>,
): PackageEndpoint {
  const endpoint = requireObject(value, label)
  if (endpoint["kind"] === "local") {
    requireKeys(endpoint, ["kind", "path"], label)
    const path = requireString(endpoint["path"], `${label}.path`)
    if (!locals.has(path)) {
      throw new Error(`${label}.path ${JSON.stringify(path)} does not name a graph node`)
    }
    return { kind: "local", path }
  }
  if (endpoint["kind"] === "registry") {
    requireKeys(endpoint, ["installPath", "kind"], label)
    const installPath = requireString(endpoint["installPath"], `${label}.installPath`)
    if (!registries.has(installPath)) {
      throw new Error(`${label}.installPath ${JSON.stringify(installPath)} does not name a graph node`)
    }
    return { installPath, kind: "registry" }
  }
  throw new Error(`${label}.kind is unsupported`)
}

function validateLocalPackage(local: LocalPackage, root: boolean): void {
  if (root) {
    if (local.path !== ".") {
      throw new Error(`package graph root path must be .`)
    }
    if (local.viewKey !== null) {
      throw new Error("package graph root viewKey must be null")
    }
  } else {
    if (local.path === ".") {
      throw new Error("only the first local package may use root path .")
    }
    validatePackagePath(local.path, programMountPath, true, "package graph local path")
    if (local.viewKey === null) {
      throw new Error("package graph non-root viewKey must be a string")
    }
    const expected = localPackageViewKey(local.path)
    if (local.viewKey !== expected) {
      throw new Error(`package graph viewKey must be ${expected}`)
    }
  }
  if (local.name !== null) validatePackageName(local.name, "package graph local name")
  if (local.version !== null) validatePackageVersion(local.version, "package graph local version")
}

function validateRegistryPackage(registry: RegistryPackage): void {
  validatePackagePath(
    registry.installPath,
    dependencyMountPath,
    false,
    "package graph registry installPath",
  )
  validatePackageIntegrity(registry.integrity)
  validatePackageName(registry.name, "package graph registry name")
  validatePackageVersion(registry.version, "package graph registry version")
}

function validatePackageName(name: string, label: string): void {
  const length = textEncoder.encode(name).length
  if (length === 0 || length > maxPackageNameBytes || !packageNamePattern.test(name)) {
    throw new Error(`${label} ${JSON.stringify(name)} is outside the exact package-name domain`)
  }
}

function validatePackageVersion(version: string, label: string): void {
  const length = textEncoder.encode(version).length
  if (length === 0 || length > maxPackageVersionBytes) {
    throw new Error(`${label} ${JSON.stringify(version)} is outside the exact package-version domain`)
  }
}

function validatePackageIntegrity(integrity: string): void {
  if (!sha512IntegrityPattern.test(integrity)) {
    throw new Error("package graph integrity is not a canonical SHA-512 SRI value")
  }
  const encoded = integrity.slice("sha512-".length)
  const digest = Buffer.from(encoded, "base64")
  if (digest.length !== 64 || digest.toString("base64") !== encoded) {
    throw new Error("package graph integrity is not a canonical SHA-512 SRI value")
  }
}

function validatePackagePath(path: string, mountPath: string, local: boolean, label: string): void {
  if (path.length === 0 || path.startsWith("/") || path.includes("\\") || /[\p{Cc}]/u.test(path)) {
    throw new Error(`${label} ${JSON.stringify(path)} is not a confined relative POSIX path`)
  }
  const components = path.split("/")
  if (
    components.some(
      (component) =>
        component === "" ||
        component === "." ||
        component === ".." ||
        textEncoder.encode(component).length > maxPackagePathComponentBytes,
    )
  ) {
    throw new Error(`${label} ${JSON.stringify(path)} is not normalized or exceeds a component bound`)
  }
  const root = components[0]!
  if (local && (root === "helmr" || root === ".helmr" || root === "node_modules")) {
    throw new Error(`${label} ${JSON.stringify(path)} uses reserved Deployment root ${JSON.stringify(root)}`)
  }
  if (!local && root === ".helmr") {
    throw new Error(`${label} ${JSON.stringify(path)} uses the reserved dependency root`)
  }
  if (textEncoder.encode(`${mountPath}/${path}\0`).length > maxMountedPackagePathBytes) {
    throw new Error(`${label} ${JSON.stringify(path)} exceeds the mounted path bound`)
  }
}

function localPackageViewKey(path: string): string {
  return createHash("sha256")
    .update(localPackageViewKeyDomain)
    .update(new Uint8Array([0]))
    .update(path)
    .digest("hex")
}

function compareResolutions(left: PackageResolution, right: PackageResolution): number {
  return (
    compareEndpoints(left.from, right.from) ||
    compareUtf8(left.dependency, right.dependency) ||
    relationshipOrder(left.relationship) - relationshipOrder(right.relationship) ||
    compareEndpoints(left.to, right.to)
  )
}

function compareEndpoints(left: PackageEndpoint, right: PackageEndpoint): number {
  return kindOrder(left.kind) - kindOrder(right.kind) || compareUtf8(locator(left), locator(right))
}

function locator(endpoint: PackageEndpoint): string {
  return endpoint.kind === "local" ? endpoint.path : endpoint.installPath
}

function kindOrder(kind: PackageKind): number {
  return kind === "local" ? 0 : 1
}

function relationshipOrder(relationship: PackageRelationship): number {
  switch (relationship) {
    case "production":
      return 0
    case "optional":
      return 1
    case "peer":
      return 2
  }
}

function compareUtf8(left: string, right: string): number {
  const leftBytes = textEncoder.encode(left)
  const rightBytes = textEncoder.encode(right)
  const length = Math.min(leftBytes.length, rightBytes.length)
  for (let index = 0; index < length; index++) {
    const difference = leftBytes[index]! - rightBytes[index]!
    if (difference !== 0) return difference
  }
  return leftBytes.length - rightBytes.length
}

function requireObject(value: JsonValue | undefined, label: string): JsonObject {
  if (value === null || typeof value !== "object" || Array.isArray(value)) {
    throw new Error(`${label} must be an object`)
  }
  return value as JsonObject
}

function requireArray(value: JsonValue | undefined, label: string): readonly JsonValue[] {
  if (!Array.isArray(value)) {
    throw new Error(`${label} must be an array`)
  }
  return value
}

function requireString(value: JsonValue | undefined, label: string): string {
  if (typeof value !== "string") {
    throw new Error(`${label} must be a string`)
  }
  return value
}

function requireNullableString(value: JsonValue | undefined, label: string): string | null {
  if (value !== null && typeof value !== "string") {
    throw new Error(`${label} must be a string or null`)
  }
  return value
}

function requireDigest(value: JsonValue | undefined, label: string): string {
  const digest = requireString(value, label)
  if (!sha256DigestPattern.test(digest)) {
    throw new Error(`${label} must be a lowercase SHA-256 digest`)
  }
  return digest
}

function requireKeys(value: JsonObject, expected: readonly string[], label: string): void {
  const keys = Object.keys(value).sort()
  if (keys.length !== expected.length || keys.some((key, index) => key !== expected[index])) {
    throw new Error(`${label} has unknown or missing members`)
  }
}

function bytesEqual(left: Uint8Array, right: Uint8Array): boolean {
  return left.length === right.length && left.every((value, index) => value === right[index])
}

File: internal/deployment/package_graph.go
package deployment

import (
	"bytes"
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"regexp"
	"strings"
	"unicode"
	"unicode/utf8"

	"github.com/helmrdotdev/helmr/internal/jsoncanon"
)

const (
	PackageGraphFormatVersion = 0

	PackageKindLocal    = PackageKind("local")
	PackageKindRegistry = PackageKind("registry")

	PackageRelationshipProduction = PackageRelationship("production")
	PackageRelationshipOptional   = PackageRelationship("optional")
	PackageRelationshipPeer       = PackageRelationship("peer")

	maxPackageNameBytes       = 214
	maxPackageVersionBytes    = 255
	maxPackagePathComponent   = 255
	maxMountedPackagePath     = 4096
	sha512DigestBytes         = 64
	localPackageViewKeyDomain = "helmr.local-package-view.v0"
	programMountPath          = "/opt/helmr/program"
	dependencyMountPath       = "/opt/helmr/program/node_modules"
)

var packageNamePattern = regexp.MustCompile(`^(?:[a-z0-9][a-z0-9._~-]*|@[a-z0-9][a-z0-9._~-]*/[a-z0-9][a-z0-9._~-]*)$`)

type PackageKind string

type PackageRelationship string

type LocalPackage struct {
	ManifestDigest string  `json:"manifestDigest"`
	Name           *string `json:"name"`
	Path           string  `json:"path"`
	Version        *string `json:"version"`
	ViewKey        *string `json:"viewKey"`
}

type RegistryPackage struct {
	InstallPath string `json:"installPath"`
	Integrity   string `json:"integrity"`
	Name        string `json:"name"`
	Version     string `json:"version"`
}

type PackageEndpoint struct {
	InstallPath *string     `json:"installPath,omitempty"`
	Kind        PackageKind `json:"kind"`
	Path        *string     `json:"path,omitempty"`
}

type PackageResolution struct {
	Dependency   string              `json:"dependency"`
	From         PackageEndpoint     `json:"from"`
	Relationship PackageRelationship `json:"relationship"`
	To           PackageEndpoint     `json:"to"`
}

type PackageGraph struct {
	FormatVersion    int                 `json:"formatVersion"`
	LocalPackages    []LocalPackage      `json:"localPackages"`
	RegistryPackages []RegistryPackage   `json:"registryPackages"`
	Resolutions      []PackageResolution `json:"resolutions"`
}

func ParsePackageGraph(raw []byte) (PackageGraph, error) {
	if len(raw) == 0 || len(raw) > int(maxProgramFileSizeBytes) {
		return PackageGraph{}, fmt.Errorf("package graph size is outside [1,%d]", maxProgramFileSizeBytes)
	}
	canonical, err := jsoncanon.Transform(raw)
	if err != nil {
		return PackageGraph{}, fmt.Errorf("canonicalize package graph: %w", err)
	}
	if !bytes.Equal(raw, canonical) {
		return PackageGraph{}, fmt.Errorf("package graph is not RFC 8785 canonical JSON")
	}

	var graph PackageGraph
	decoder := json.NewDecoder(bytes.NewReader(raw))
	decoder.DisallowUnknownFields()
	if err := decoder.Decode(&graph); err != nil {
		return PackageGraph{}, fmt.Errorf("decode package graph: %w", err)
	}
	if err := ensureEOF(decoder, "package graph"); err != nil {
		return PackageGraph{}, err
	}
	if err := ValidatePackageGraph(graph); err != nil {
		return PackageGraph{}, err
	}
	complete, err := CanonicalPackageGraph(graph)
	if err != nil {
		return PackageGraph{}, err
	}
	if !bytes.Equal(raw, complete) {
		return PackageGraph{}, fmt.Errorf("package graph does not match the complete canonical v0 shape")
	}
	return graph, nil
}

func CanonicalPackageGraph(graph PackageGraph) ([]byte, error) {
	if err := ValidatePackageGraph(graph); err != nil {
		return nil, err
	}
	raw, err := json.Marshal(graph)
	if err != nil {
		return nil, fmt.Errorf("encode package graph: %w", err)
	}
	canonical, err := jsoncanon.Transform(raw)
	if err != nil {
		return nil, fmt.Errorf("canonicalize package graph: %w", err)
	}
	if len(canonical) > int(maxProgramFileSizeBytes) {
		return nil, fmt.Errorf("package graph size is outside [1,%d]", maxProgramFileSizeBytes)
	}
	return canonical, nil
}

func ValidatePackageGraph(graph PackageGraph) error {
	if graph.FormatVersion != PackageGraphFormatVersion {
		return fmt.Errorf("package graph formatVersion = %d, want %d", graph.FormatVersion, PackageGraphFormatVersion)
	}
	if graph.LocalPackages == nil {
		return fmt.Errorf("package graph localPackages must be an array")
	}
	if graph.RegistryPackages == nil {
		return fmt.Errorf("package graph registryPackages must be an array")
	}
	if graph.Resolutions == nil {
		return fmt.Errorf("package graph resolutions must be an array")
	}
	if len(graph.LocalPackages) == 0 || graph.LocalPackages[0].Path != "." {
		return fmt.Errorf("package graph localPackages must begin with exactly one root record")
	}

	locals := make(map[string]LocalPackage, len(graph.LocalPackages))
	localNames := make(map[string]struct{}, len(graph.LocalPackages))
	viewKeys := make(map[string]struct{}, len(graph.LocalPackages))
	for position, local := range graph.LocalPackages {
		if err := validateLocalPackage(local, position == 0); err != nil {
			return fmt.Errorf("package graph localPackages[%d]: %w", position, err)
		}
		if _, exists := locals[local.Path]; exists {
			return fmt.Errorf("package graph localPackages contains duplicate path %q", local.Path)
		}
		locals[local.Path] = local
		if local.Name != nil {
			if _, exists := localNames[*local.Name]; exists {
				return fmt.Errorf("package graph localPackages contains ambiguous name %q", *local.Name)
			}
			localNames[*local.Name] = struct{}{}
		}
		if local.ViewKey != nil {
			if _, exists := viewKeys[*local.ViewKey]; exists {
				return fmt.Errorf("package graph localPackages contains colliding viewKey %q", *local.ViewKey)
			}
			viewKeys[*local.ViewKey] = struct{}{}
		}
		if position > 1 {
			previous := graph.LocalPackages[position-1].Path
			if bytes.Compare([]byte(previous), []byte(local.Path)) >= 0 {
				return fmt.Errorf("package graph localPackages are not in canonical path order at position %d", position)
			}
		}
		for separator := strings.IndexByte(local.Path, '/'); separator >= 0; {
			ancestor := local.Path[:separator]
			if _, exists := locals[ancestor]; exists {
				return fmt.Errorf("package graph non-root local paths %q and %q overlap", ancestor, local.Path)
			}
			next := strings.IndexByte(local.Path[separator+1:], '/')
			if next < 0 {
				break
			}
			separator += next + 1
		}
	}

	registries := make(map[string]struct{}, len(graph.RegistryPackages))
	for position, registry := range graph.RegistryPackages {
		if err := validateRegistryPackage(registry); err != nil {
			return fmt.Errorf("package graph registryPackages[%d]: %w", position, err)
		}
		if _, exists := registries[registry.InstallPath]; exists {
			return fmt.Errorf("package graph registryPackages contains duplicate installPath %q", registry.InstallPath)
		}
		registries[registry.InstallPath] = struct{}{}
		if position > 0 && bytes.Compare(
			[]byte(graph.RegistryPackages[position-1].InstallPath),
			[]byte(registry.InstallPath),
		) >= 0 {
			return fmt.Errorf("package graph registryPackages are not in canonical installPath order at position %d", position)
		}
	}

	for position, resolution := range graph.Resolutions {
		if err := validatePackageResolution(resolution, locals, registries); err != nil {
			return fmt.Errorf("package graph resolutions[%d]: %w", position, err)
		}
		if position > 0 && comparePackageResolutions(graph.Resolutions[position-1], resolution) >= 0 {
			return fmt.Errorf("package graph resolutions are not in canonical order at position %d", position)
		}
	}
	return nil
}

func validateLocalPackage(local LocalPackage, root bool) error {
	if !sha256DigestPattern.MatchString(local.ManifestDigest) {
		return fmt.Errorf("manifestDigest is not a lowercase SHA-256 digest")
	}
	if root {
		if local.Path != "." {
			return fmt.Errorf("root path = %q, want .", local.Path)
		}
		if local.ViewKey != nil {
			return fmt.Errorf("root viewKey must be null")
		}
	} else {
		if local.Path == "." {
			return fmt.Errorf("only the first local package may use root path .")
		}
		if err := validatePackagePath(local.Path, programMountPath, true); err != nil {
			return fmt.Errorf("path: %w", err)
		}
		if local.ViewKey == nil {
			return fmt.Errorf("non-root viewKey must be a string")
		}
		want := localPackageViewKey(local.Path)
		if *local.ViewKey != want {
			return fmt.Errorf("viewKey = %q, want %q", *local.ViewKey, want)
		}
	}
	if local.Name != nil {
		if err := validatePackageName(*local.Name); err != nil {
			return fmt.Errorf("name: %w", err)
		}
	}
	if local.Version != nil {
		if err := validatePackageVersion(*local.Version); err != nil {
			return fmt.Errorf("version: %w", err)
		}
	}
	return nil
}

func validateRegistryPackage(registry RegistryPackage) error {
	if err := validatePackagePath(registry.InstallPath, dependencyMountPath, false); err != nil {
		return fmt.Errorf("installPath: %w", err)
	}
	if err := validatePackageIntegrity(registry.Integrity); err != nil {
		return err
	}
	if err := validatePackageName(registry.Name); err != nil {
		return fmt.Errorf("name: %w", err)
	}
	if err := validatePackageVersion(registry.Version); err != nil {
		return fmt.Errorf("version: %w", err)
	}
	return nil
}

func validatePackageResolution(
	resolution PackageResolution,
	locals map[string]LocalPackage,
	registries map[string]struct{},
) error {
	if err := validatePackageName(resolution.Dependency); err != nil {
		return fmt.Errorf("dependency: %w", err)
	}
	switch resolution.Relationship {
	case PackageRelationshipProduction, PackageRelationshipOptional, PackageRelationshipPeer:
	default:
		return fmt.Errorf("relationship %q is unsupported", resolution.Relationship)
	}
	if err := validatePackageEndpoint(resolution.From, locals, registries); err != nil {
		return fmt.Errorf("from: %w", err)
	}
	if err := validatePackageEndpoint(resolution.To, locals, registries); err != nil {
		return fmt.Errorf("to: %w", err)
	}
	if resolution.To.Kind == PackageKindLocal {
		if resolution.From.Kind != PackageKindLocal {
			return fmt.Errorf("registry-to-local resolution is unsupported")
		}
		local := locals[*resolution.To.Path]
		if local.Name == nil {
			return fmt.Errorf("local target %q has no name", *resolution.To.Path)
		}
		if resolution.Dependency != *local.Name {
			return fmt.Errorf("dependency %q does not equal local target name %q", resolution.Dependency, *local.Name)
		}
	}
	return nil
}

func validatePackageEndpoint(
	endpoint PackageEndpoint,
	locals map[string]LocalPackage,
	registries map[string]struct{},
) error {
	switch endpoint.Kind {
	case PackageKindLocal:
		if endpoint.Path == nil || endpoint.InstallPath != nil {
			return fmt.Errorf("local endpoint must contain exactly kind and path")
		}
		if _, exists := locals[*endpoint.Path]; !exists {
			return fmt.Errorf("local path %q does not name a graph node", *endpoint.Path)
		}
	case PackageKindRegistry:
		if endpoint.InstallPath == nil || endpoint.Path != nil {
			return fmt.Errorf("registry endpoint must contain exactly installPath and kind")
		}
		if _, exists := registries[*endpoint.InstallPath]; !exists {
			return fmt.Errorf("registry installPath %q does not name a graph node", *endpoint.InstallPath)
		}
	default:
		return fmt.Errorf("kind %q is unsupported", endpoint.Kind)
	}
	return nil
}

func validatePackageName(name string) error {
	if len(name) == 0 || len(name) > maxPackageNameBytes || !packageNamePattern.MatchString(name) {
		return fmt.Errorf("%q is outside the exact package-name domain", name)
	}
	return nil
}

func validatePackageVersion(version string) error {
	if len(version) == 0 || len(version) > maxPackageVersionBytes || !utf8.ValidString(version) {
		return fmt.Errorf("%q is outside the exact package-version domain", version)
	}
	return nil
}

func validatePackageIntegrity(integrity string) error {
	const prefix = "sha512-"
	if !strings.HasPrefix(integrity, prefix) {
		return fmt.Errorf("integrity is not a canonical SHA-512 SRI value")
	}
	encoded := strings.TrimPrefix(integrity, prefix)
	digest, err := base64.StdEncoding.DecodeString(encoded)
	if err != nil || len(digest) != sha512DigestBytes || base64.StdEncoding.EncodeToString(digest) != encoded {
		return fmt.Errorf("integrity is not a canonical SHA-512 SRI value")
	}
	return nil
}

func validatePackagePath(path, mountPath string, local bool) error {
	if path == "" || !utf8.ValidString(path) || strings.HasPrefix(path, "/") || strings.Contains(path, "\\") {
		return fmt.Errorf("%q is not a confined relative POSIX path", path)
	}
	for _, character := range path {
		if unicode.IsControl(character) {
			return fmt.Errorf("%q contains a control character", path)
		}
	}
	components := strings.Split(path, "/")
	for _, component := range components {
		if component == "" || component == "." || component == ".." {
			return fmt.Errorf("%q is not normalized", path)
		}
		if len(component) > maxPackagePathComponent {
			return fmt.Errorf("%q contains a component over %d bytes", path, maxPackagePathComponent)
		}
	}
	root := components[0]
	if local && (root == "helmr" || root == ".helmr" || root == "node_modules") {
		return fmt.Errorf("%q uses reserved Deployment root %q", path, root)
	}
	if !local && root == ".helmr" {
		return fmt.Errorf("%q uses reserved dependency root %q", path, root)
	}
	if len(mountPath)+1+len(path)+1 > maxMountedPackagePath {
		return fmt.Errorf("%q exceeds the mounted path bound", path)
	}
	return nil
}

func localPackageViewKey(path string) string {
	hash := sha256.New()
	hash.Write([]byte(localPackageViewKeyDomain))
	hash.Write([]byte{0})
	hash.Write([]byte(path))
	return hex.EncodeToString(hash.Sum(nil))
}

func comparePackageResolutions(left, right PackageResolution) int {
	if comparison := comparePackageEndpoints(left.From, right.From); comparison != 0 {
		return comparison
	}
	if comparison := bytes.Compare([]byte(left.Dependency), []byte(right.Dependency)); comparison != 0 {
		return comparison
	}
	if comparison := compareInts(
		packageRelationshipOrder(left.Relationship),
		packageRelationshipOrder(right.Relationship),
	); comparison != 0 {
		return comparison
	}
	return comparePackageEndpoints(left.To, right.To)
}

func comparePackageEndpoints(left, right PackageEndpoint) int {
	if comparison := compareInts(packageKindOrder(left.Kind), packageKindOrder(right.Kind)); comparison != 0 {
		return comparison
	}
	return bytes.Compare([]byte(packageEndpointLocator(left)), []byte(packageEndpointLocator(right)))
}

func packageEndpointLocator(endpoint PackageEndpoint) string {
	if endpoint.Kind == PackageKindLocal && endpoint.Path != nil {
		return *endpoint.Path
	}
	if endpoint.Kind == PackageKindRegistry && endpoint.InstallPath != nil {
		return *endpoint.InstallPath
	}
	return ""
}

func packageKindOrder(kind PackageKind) int {
	if kind == PackageKindLocal {
		return 0
	}
	if kind == PackageKindRegistry {
		return 1
	}
	return 2
}

func packageRelationshipOrder(relationship PackageRelationship) int {
	switch relationship {
	case PackageRelationshipProduction:
		return 0
	case PackageRelationshipOptional:
		return 1
	case PackageRelationshipPeer:
		return 2
	default:
		return 3
	}
}

func compareInts(left, right int) int {
	switch {
	case left < right:
		return -1
	case left > right:
		return 1
	default:
		return 0
	}
}
